geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java

```diff
@@ -42,16 +45,65 @@ public RedisList() {
     this.elementList = new SizeableByteArrayList();
   }
 
+  /**
+   * @param start start index of desired elments
+   * @param stop stop index of desired elments
+   * @return list of elements in the range (inclusive).
+   */
+  public List<byte[]> lrange(int start, int stop) {
+    start = transformNegIndexToPosIndex(start);
+    stop = transformNegIndexToPosIndex(stop);
+
+    // Out of range negative index changes to 0
+    start = Math.max(0, start);
+
+    int elementSize = elementList.size();
+    if (start > stop || elementSize <= start) {
+      return Collections.emptyList();
+    }
+
+    // Out of range positive stop index changes to last index available
+    stop = elementSize <= stop ? elementSize - 1 : stop;
+
+    List<byte[]> result = new LinkedList<>();
+    int curIndex;
+
+    // Finds the shortest distance to access nodes in range
+    if (start <= elementSize - stop - 1) {
+      // Starts at head to access nodes at start index then iterates forwards
+      curIndex = 0;
+      ListIterator<byte[]> iterator = elementList.listIterator(curIndex);
+
+      while (iterator.hasNext() && curIndex <= stop) {
+        byte[] element = iterator.next();
+        if (start <= curIndex) {
+          result.add(element);
+        }
+        curIndex = iterator.nextIndex();
+      }
+    } else {
+      // Starts at tail to access nodes at stop index then iterates backwards
+      curIndex = elementSize - 1;
+      ListIterator<byte[]> iterator = elementList.listIterator(elementSize);
+
+      while (iterator.hasPrevious() && start <= curIndex) {
+        byte[] element = iterator.previous();
+        if (curIndex <= stop) {
+          result.add(0, element); // LinkedList is used to add to head
+        }
+        curIndex = iterator.previousIndex();
+      }
+    }
+    return result;
+  }
+
   /**
    * @param index index of desired element. Positive index starts at the head. Negative index starts
    *        at the tail.
    * @return element at index. Null if index is out of range.
    */
   public byte[] lindex(int index) {
-    if (index < 0) {
-      // Changes negative index to corresponding positive index to utilize get(int index)
-      index = elementList.size() + index;
-    }
+    index = transformNegIndexToPosIndex(index);
 
     if (index < 0 || elementList.size() <= index) {
```

Comment:
if transformNegIndexToPosIndex checked to see if it was going to return something < 0 and instead returns Integer.MAX then this if statement could drop the "index < 0" check.
Maybe not Integer.MAX just so we could support lists of size 2G. Are redis native lists also limited to <= 2G elements?
But we could just have this method make sure the index it returns would be a valid list index [0...size-1]. If not then it could either throw an exception or return an exceptional value (for example we could define a constant int INVALID_INDEX = -1). Then the callers can just ask if the result is invalid and then return null/emptyList. Once lrange knows start and stop are both valid it just needs the additional check that start is not greator that stop.